File: scala/app/src/command/Ping.scala
package command

import error.CommandExecutionError
import resp.Data

extension (c: Command)
  def ping(): Either[CommandExecutionError, Data] = {
    val message = c.arguments.nextOption().getOrElse(Data.BulkString("PONG"))

    // Ensure that no more arguments are provided.
    if(c.arguments.hasNext) {
      return Left(CommandExecutionError.WrongArguments)
    }

    Right(message)
  }

File: typescript/src/command/ping.ts
import { Effect } from "effect"
import { Command } from "./command"
import { WrongArguments, type CommandExecutionError } from "../error"
import { BulkString, type TData } from "../resp/data"

export class PingCommand extends Command {
  override handle( ): Effect.Effect<TData, CommandExecutionError> {
    const messageOption = this.args.next( )

    const message = !messageOption.done
      ? messageOption.value
      : BulkString({ string: "PONG" })

    // Ensure that no more arguments are provided.
    if(!this.args.next( ).done)
      return Effect.fail(WrongArguments( ))

    return Effect.succeed(message)
  }
}
